enhancements/agent-installer/automated-workflow.md

```diff
@@ -0,0 +1,491 @@
+format and approach first. Building minimal ISOs with separate root
+filesystems or images that can be used with hosts that PXE boot will
+come later.
+
+We propose to leave up to the user's orchestration tool the
+responsibilities of managing and storing the images, serving them on
+the network, and especially configuring hosts to boot them. This gives
+users the flexibility to perform those operations in any way they
+like, with whatever tool they prefer. A design for an all encompassing
+approach based on tools provided entirely by Red Hat is left for a
+later enhancement.
+
+The `create image` command will need to take inputs similar to what
+the baremetal IPI installer or ZTP APIs use to describe the cluster
+(name, API IPs, etc.) and host-specific network settings (static IPs,
+VLANs, etc.). Using the existing file and custom resource types allows
+us to use the same version management techniques that we do for the
+APIs within a cluster.
+
+The `create image` command will produce a single image as output, to
+save storage space and reduce management complexity. The assisted
+installer service can already generate images with boot scripts that
+recognize which host is booting and apply host-specific network
+settings and roles (such as running different services) accordingly,
+and we will reuse that technique (and code).
+
+One host, designated as "node0", will be configured to run the
+assisted service and a client that uses the assisted service API to
+automate the deployment.
+
+All other hosts will run the assisted agent, configured to talk to the
+service on node0.
+
+When all of the images are booted, the client on node0 will use the
+assisted service to trigger the deployment of the cluster (including
+applicable pre-flight validations) without further interaction from
+the user.
+
+The `create image` command will produce as output credentials to
+access the cluster that is about to be created. This allows a client,
+such as a tool monitoring deployment progress, to use the cluster's
+API as soon as the API service is available.
+
+During deployment, the third-party orchestration tool will need a way
+to monitor progress and discover errors. The assisted service
+read-only API endpoints running on node0 will report that data. In the
+event of failure, a read-only endpoint in that service will be
+available to download any logs that have been collected.
+
+We will be taking advantage of the assisted service's ability to run
+the bootstrap services "in place" to avoid having a separate bootstrap
+VM. Bootstrap will run on node0, where the assisted service runs, to
+ensure that the service and its API are available for as long as
+possible. When bootstrapping completes and node0 reboots, progress
+monitoring will need to shift to the OpenShift APIs (`ClusterVersion`
+and `ClusterOperator`). We will extend the `openshift-install wait-for
+install-complete` command to recognize when the cluster being deployed
```

Comment:
I'm not sure that anything needs to be extended here. I think it would just work.

Reply:
I'm also a bit lost here.  Once node0 reboots, there are no Assisted components running anywhere (agents or service).  That means the logs/progress API endpoints described in the previous paragraph are no longer available.  In the paradigm where the service runs externally to the cluster, we do have a poorly-named component ("assisted controller", which is not a controller) that runs on the installed cluster and reports back to the service on progress and logs.
Now, in case of installation failure in the final steps, the user would need to understand that those REST endpoints disappear and they now need to use k8s APIs for further monitoring.  We need to be conscious of trading off this UX degradation with running the Assisted service externally to the cluster (on a laptop or single host that is used for all installations for the site).  We also need to remember that the Assisted Service currently provides credentials to the installed cluster, which are generated when the user clicks "install".

Reply:
Also, the users will have no access to logs post pivot.

Reply:
Just to be clear, the poorly-named component in the new cluster is a Job that completes eventually and goes away. So it's not a long-running component in the new cluster.

I think the point here is that it will get confused if it can't connect to the service, which will be the case when node0 reboots and starts joining the cluster. But its role in reporting on progress and errors prior to that pivot is probably still valuable (or maybe even required), so some work will be needed to adjust assumptions in one place or another to fit this new workflow.

Reply:
https://github.com/openshift/enhancements/pull/1067/files#r832051897

Reply:
Requiring people to use openshift-install to do any monitoring or debugging is a serious usability drawback.

Reply:
> Requiring people to use openshift-install to do any monitoring or debugging is a serious usability drawback.

I feel like this is a response to my question, but I do not understand it or how it applies to my question. Can you elaborate?

Reply:
I've replaced `wait-for install-complete` with `agent wait-for`.

@avishayt , the user is not required to use the CLI to monitor progress. They can still use the REST API on node0 themselves, if that's what they prefer.